// luna-console/cjs/share/Component.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Emitter_1 = __importDefault(require("licia/Emitter"));
const _1 = __importDefault(require("licia/$"));
const util_1 = require("./util");
const each_1 = __importDefault(require("licia/each"));
class Component extends Emitter_1.default {
    constructor(container, { compName }) {
        super();
        this.compName = compName;
        this.c = util_1.classPrefix(compName);
        this.options = {};
        this.container = container;
        this.$container = _1.default(container);
        this.$container.addClass(`luna-${compName}`);
    }
    destroy() {
        this.$container.rmClass(`luna-${this.compName}`);
        this.$container.html('');
        this.emit('destroy');
        this.removeAllListeners();
    }
    setOption(name, val) {
        const options = this.options;
        let newOptions = {};
        if (typeof name === 'string') {
            newOptions[name] = val;
        }
        else {
            newOptions = name;
        }
        each_1.default(newOptions, (val, name) => {
            const oldVal = options[name];
            options[name] = val;
            this.emit('optionChange', name, val, oldVal);
        });
    }
    find(selector) {
        return this.$container.find(this.c(selector));
    }
}
exports.default = Component;
